### ISSUE TITLE
How do you add code coverage metrics

### ISSUE THREAD

**Issue by Swoorup** [tcplugins/tcWebHooks#145]
Is it possible to get the code coverage figures once the build finishes? i.e the figures in the Reported Statistics Values? 

I have searched through templates and examples but it doesn't seem possible?

![image](https://user-images.githubusercontent.com/3408009/62102445-8cb09a00-b2cb-11e9-940c-c043c2537882.png)

**Comment by netwolfuk** [comment]
Yes. You're right. That data is not available to the templating engine.

Once it's added it would be accessible via methods available to the velocity templates.

I just need to come up with a better way to add it to the Velocity context. Adding it the existing payload would probably break the other formats because they would try the serialise it and most likely fail to do so.

**Comment by netwolfuk** [comment]
Depends on #31

**Comment by netwolfuk** [comment]
Hi @Swoorup  I have added preliminary support for accessing the build object (SBuild) via Velocity.

Please [try a build from this branch,](https://teamcity.jetbrains.com/buildConfiguration/WebHooksAndOtherPlugins_TcWebHooks?branch=issue_31-add_more_items_to_velocity_payload&buildTypeTab=overview) and create a template using the example on the [Velocity page on the wiki](https://github.com/tcplugins/tcWebHooks/wiki/WebHook-Templates-:-The-Velocity-Templating-Engine#print-the-builds-statistics).

This is what I get in my testing....

Example template:
```
#set ($buildStatistics = $build.fullStatistics)
#set ($statsValues = $build.statisticValues)
{
    "buildName": "${buildType.name}",
    "projectName" : "${project.externalId}",
    "successfulTests": $buildStatistics.passedTestCount, 
    "failedTests" : $buildStatistics.failedTestCount,
    "ignoredTests": $buildStatistics.ignoredTestCount,
    "mutedTests"  : $buildStatistics.mutedTestsCount,
#foreach ($test in $buildStatistics.allTests)
    #if( $test.status.isIgnored() )
       "test-$test.testRunId" : "$test.test.name - $test.status",
    #end
#end
#foreach ($mapEntry in $statsValues.entrySet())
  #if ( $mapEntry.value )
    "$mapEntry.key": $mapEntry.value.intValue(),
  #end
#end
    "buildResult" : "${buildResult}"
}
```

Example Output:
```

  "buildName": "tcWebHooks - GitHub",
  "projectName": "TcPlugins",
  "successfulTests": 338,
  "failedTests": 0,
  "ignoredTests": 9,
  "mutedTests": 0,
  "test-1640": "webhook.teamcity.reporting.ReportPayloadFactoryImplTest.testBuildReportPayload - UNKNOWN",
  "test-5617": "webhook.teamcity.WebHookContentBuilderTest.testBuildWebHookContent - UNKNOWN",
  "test-6066": "webhook.teamcity.WebHookListenerTest.testBuildChangedStatusSRunningBuildStatusStatus - UNKNOWN",
  "test-6069": "webhook.teamcity.WebHookListenerTest.testBuildStartedSRunningBuild - UNKNOWN",
  "test-6072": "webhook.teamcity.WebHookListenerTest.testBeforeBuildFinishSRunningBuild - UNKNOWN",
  "test-6077": "webhook.teamcity.WebHookListenerTest.testBuildInterruptedSRunningBuild - UNKNOWN",
  "test-6082": "webhook.teamcity.WebHookListenerTest.testBuildFinishedSRunningBuildSuccessAfterSuccess - UNKNOWN",
  "test-6085": "webhook.teamcity.WebHookListenerTest.testBuildFinishedSRunningBuild - UNKNOWN",
  "test-6088": "webhook.teamcity.WebHookListenerTest.testBuildFinishedSRunningBuildSuccessAfterFailure - UNKNOWN",
  "buildStageDuration:artifactsPublishing": 2889,
  "CodeCoverageM": 53,
  "CodeCoverageR": 31,
  "JacocoComplexityTotal": 6126,
  "CodeCoverageL": 67,
  "ArtifactsSize": 5385647,
  "PassedTestCount": 338,
  "CodeCoverageAbsMTotal": 3879,
  "buildStageDuration:buildFinishing": 50,
  "CodeCoverageB": 62,
  "buildStageDuration:buildStepRUNNER_13": 74789,
  "JacocoComplexityCovered": 2512,
  "CodeCoverageC": 83,
  "BuildDuration": 240812,
  "VisibleArtifactsSize": 17871,
  "IgnoredTestCount": 9,
  "buildStageDuration:buildStepRUNNER_4": 160752,
  "TotalTestCount": 347,
  "CodeCoverageAbsRTotal": 4483,
  "serverSideBuildFinishing": 49,
  "CodeCoverageAbsLTotal": 12644,
  "CodeCoverageAbsLCovered": 8507,
  "CodeCoverageAbsCTotal": 437,
  "buildStageDuration:sourcesUpdate": 2071,
  "CodeCoverageAbsMCovered": 2067,
  "BuildTestStatus": 1,
  "SuccessRate": 1,
  "CodeCoverageAbsBCovered": 43818,
  "CodeCoverageAbsCCovered": 366,
  "CodeCoverageAbsRCovered": 1404,
  "buildStageDuration:toolsUpdating": 1,
  "queueWaitReason:Build_settings_have_not_been_finalized": 44,
  "buildStageDuration:firstStepPreparation": 3,
  "CodeCoverageAbsBTotal": 70412,
  "TimeSpentInQueue": 40,
  "BuildDurationNetTime": 235541,
  "buildResult": "success"
```

**Comment by Swoorup** [comment]
That's pretty cool. Can confirm that it works. 👍 
Can't wait for the release

**Comment by netwolfuk** [comment]
Will you wait for a release that's not alpha (no ETA yet) or do you mean when that branch is merged to master and `1.2.0alpha-3` is released?

**Comment by Swoorup** [comment]
I prefer to wait for a release that's not alpha. 

Have a couple of teamcity slack build messages that we could replace via this plugins which opens to bit more customization.

I resolved the initial code coverage by instead using a custom powershell script as a build step which is another build project and has a dependency on the project on which I want to report coverage metrics from (not really ideal). 

It then can read of dependency build metrics and then outputs the coverage % as progress bar like (🔳🔳🔳🔳🔲). Probably doable via velocity templates as well, but a bit complicated.

**Comment by netwolfuk** [comment]
Ok, I'll leave this ticket open until 1.2 is released.

Yeah. That would be pretty simple in velocity

**Comment by netwolfuk** [comment]
This is available for testing in the [v1.2.0-alpha.3](https://github.com/tcplugins/tcWebHooks/releases/tag/v1.2.0-alpha.3) release.
